Add rotating log support

walac opened the pull request:
Replace the log system by Python logging builtin package. We add support
to log both on screen and on disk. On disk logging is done using a
rotating scheme.

Commit:
Add support for rotating logs on disk.

Replace the log system by Python logging builtin package. We add support
to log both on screen and on disk. On disk logging is done using a
rotating scheme.

walac commented:
r? @vdjeric

vdjeric commented on symLogging.py:
```diff
@@ -1,29 +1,54 @@
+from logging.handlers import RotatingFileHandler
 
 gEnableTracing = False
+logger = None
```
call it gLogger for consistency

vdjeric replied:
or gLog :)

vdjeric commented on symLogging.py:
```diff
@@ -1,29 +1,54 @@
+  global gEnableTracing, logger
   gEnableTracing = isEnabled
 
+  filepath = logOptions.get('logPath', '/var/log/snappy')
```
this server will be run on *nix servers and developers' Windows machines as well, so hard-coding paths like this isn't ideal. Either check for OS type and use different paths for each OS, or use tempfile.gettempdir to log to a temporary directory if no logging dir is configured in the .ini

vdjeric commented on symLogging.py:
```diff
@@ -1,29 +1,54 @@
+              filename=os.path.join(filepath, 'snappy.log'),
+              mode='a',
+              maxBytes=int(logOptions.get("maxFileSize", 1024*1024)),
+              backupCount=int(logOptions.get("maxFiles", 10)))
```
~~make these settings configurable as well~~ sorry, I misread, this is fine

vdjeric commented on symLogging.py:
```diff
@@ -1,29 +1,54 @@
+
+  streamHandler = logging.StreamHandler()
+
+  fmt = logging.Formatter('%(asctime)s:%(levelname)s:%(message)s')
```
this format might be tough to read, put a tab between the fields. tabs should still be easily parseable with awk etc

vdjeric commented on symLogging.py:
```diff
@@ -1,29 +1,54 @@
 def LogError(string):
-  threadName = threading.currentThread().getName().ljust(12)
-  print >> sys.stderr, time.asctime() + " " + threadName + " ERROR " + string
+  if logger.isEnabledFor(logging.ERROR):
```
won't the logger library do this check on its own?

vdjeric commented on symLogging.py:
```diff
@@ -1,29 +1,54 @@
-  print >> sys.stdout, time.asctime() + " " + threadName + "       " + string
+  if logger.isEnabledFor(logging.INFO):
+    threadName = threading.currentThread().getName().ljust(12)
+    logger.info("%s       %s", threadName, string)
```
why the extra spaces between the threadName and string for LogMessage?

walac replied:
Yep, but I added the `if` to avoid unnecessary computation.

vdjeric replied:
Getting the thread name shouldn't be a lot of computation, but it's your call
Btw, in this patch it's not possible to specifying the logging level so logger.isEnabledFor(logging.INFO) will always be True. We should make it possible to specify the logging level via config file or the debug console

vdjeric commented on symbolicationWebService.py:
```diff
@@ -163,8 +163,8 @@ def ReadConfigFile():
-  if set(configParser.sections()) != set(["General", "SymbolPaths", "SymbolURLs"]):
-    LogError("Config file should be made up of three sections: 'General', 'SymbolPaths' and 'SymbolURLs'")
+  if set(configParser.sections()) != set(["General", "SymbolPaths", "SymbolURLs", "Log"]):
+    print("Config file should be made up of three sections: 'General', 'SymbolPaths' and 'SymbolURLs'")
```
shouldn't this error message also warn about missing Log section?

vdjeric commented on symLogging.py:
```diff
@@ -1,29 +1,54 @@
+
 def LogTrace(string):
   global gEnableTracing
   if gEnableTracing:
```
the logging library knows its current log level, right?

walac replied:
Yes, it does. But there is the computation done to format the message string just to go into the function call and the logging library suppresses the message output.

vdjeric replied:
ok, so logger.isEnabledFor instead?

vdjeric commented on symbolicationWebService.py:
```diff
@@ -199,7 +201,7 @@ def Main():
     return 1
 
-  SetTracingEnabled(gOptions["enableTracing"] > 0)
+  SetTracingEnabled(gOptions["enableTracing"] > 0, gOptions["Log"])
```
same as before, couldn't you rely on the logging library's log-level instead of maintaining gEnableTracing

vdjeric replied:
maybe rename SetTracingEnabled to SetLoggingOptions

vdjeric commented on sample.ini:
```diff
@@ -1,29 +1,37 @@
+
+maxCacheEntries = 2000000
+
+mruSymbolStateFile = /var/snappy-mru-symbols.json
```
can you turn this into 2 sample files: sample-windows.ini and sample-linux.ini? then these config files will work out of the box for both server deployment and for Windows developers who are profiling their local builds with local symbols i.e. https://developer.mozilla.org/en-US/docs/Mozilla/Performance/Profiling_with_the_Built-in_Profiler_and_Local_Symbols_on_Windows#Profiling_local_builds_%28without_using_talos%29
and then we can use the appropriate line endings for each

walac replied:
hrm, I fear this might bring inconsistency for configuration across files. We still have a mix of CRLF/LF files in the repository. sample.ini was more problematic because its line endings had an extra LF. One way to address this is adding a [.gitattributes](https://www.kernel.org/pub/software/scm/git/docs/gitattributes.html) file with the following:

```
* text=auto
```

This will enforce files go into the repository with LFs. Windows developers can set their git configuration [core.autocrlf](https://help.github.com/articles/dealing-with-line-endings/) to:

```
git config --global core.autocrlf true
```

Then they will get CRLF for checked out files, but commits will store LF in the repo.

vdjeric replied:
sounds good!

vdjeric commented on sample.ini:
```diff
@@ -1,29 +1,37 @@
+[SymbolPaths]
 
-Firefox = /mnt/netapp/breakpad/symbols_ffx
+Firefox = /mnt/netapp/breakpad/symbols_ffx
```
i'd like to change how these symbol paths are handled.. the official symbolapi.mozilla.org server won't have access to any symbol directories when doing symbol lookups, so it doesn't make sense to check for symbols on-disk in production before checking the (future) disk cache or fetching from S3. Windows devs will need to use on-disk symbols + remoteSymbolServer only. So using S3 or local symbols should be mutually exclusive options

walac replied:
~~This one I could not find a reasonable solution. If you comment out the local paths, them snappy will not try any local search. If you comment out the symbol URL, and remove [this line](https://github.com/mozilla/Snappy-Symbolication-Server/blob/master/symbolicationWebService.py#L46), no S3 fetch happens.~~ already discussed in IRC/email

Commit:
nit: Store all text files with LF line ending format.

Windows users should do

    git config --global core.autocrlf true

to get CRLF for the checked out text files.

vdjeric commented on sample-linux.ini:
```diff
@@ -0,0 +1,45 @@
+; NOTSET
+logLevel = INFO
+
+[SymbolPaths]
```
SymbolPaths should be commented out too

vdjeric commented on sample-windows.ini:
```diff
@@ -0,0 +1,45 @@
+
+maxCacheEntries = 2000000
+
+mruSymbolStateFile = /tmp/snappy-mru-symbols.json
```
this should be a Windows path.. make sure to test the changes on Windows btw

vdjeric commented on sample-windows.ini:
```diff
@@ -0,0 +1,45 @@
+
+maxFileSize = 1048576
+
+;logPath = /tmp/log/snappy
```
put it somewhere visible.. maybe %HOME% if python supports the escapes, otherwise hard-code windows temp directory

vdjeric commented on sample-windows.ini:
```diff
@@ -0,0 +1,45 @@
+
+[SymbolPaths]
+
+Firefox = /mnt/netapp/breakpad/symbols_ffx
```
use windows paths

vdjeric commented on sample.ini:
```diff
@@ -1,29 +1,43 @@
 [General]
```
we don't need sample.ini anymore

vdjeric commented on .gitignore:
```diff
@@ -1,8 +1,10 @@
+# Exclude local config files, except sample file
+##################################################
+*.ini
+!sample.ini
```
we don't need sample.ini anymore

vdjeric commented on symLogging.py:
```diff
@@ -1,29 +1,58 @@
+
+  try:
+    rotatingHandler = RotatingFileHandler(
+                filename=os.path.join(filepath, 'snappy.log'),
```
indentation is a bit funky

vdjeric commented on symLogging.py:
```diff
@@ -1,29 +1,58 @@
-    print >> sys.stdout, time.asctime() + " " + threadName + " TRACE " + string
+    gLog.log(dbgLevel, "%s\t%s", threadName, string)
+
+def LogTrace(string):
```
nit: might as well rename this to LogDebug, we don't have enableTracing in the config file anymore

Commit:
Support environment variables in the configuration file.

It uses the ConfigParser interpolation syntax %(<var>)<type>.

Environment variables are put in the DEFAULT special section, and
calling items() for any section also returns values in DEFAULT. We then
override items() to filter out defaults.

Commit:
Add platform specific configuration files.

The Linux configuration file is for production, always look for symbols
at S3.

Windows configuration file is for Windows developers. Look for symbols
locally and at remote server.

walac commented:
New round of patches. Configuration files now support environment variables.

vdjeric commented on sample-linux.ini:
```diff
@@ -1,31 +1,45 @@
 
 ; If any symbols of interest aren't available locally (e.g. Windows DLLs), uncomment line below
-
+
```
how did we end up with extra  blank lines in this file?

walac replied:
github is nuts here, cannot detect this with `git diff`

vdjeric commented on sample-windows.ini:
```diff
@@ -0,0 +1,45 @@
+
+maxCacheEntries = 2000000
+
+mruSymbolStateFile = %(TEMP)s\snappy-mru-symbols.json
```
nit: maybe %(HOME) here to make it easier for users to find?

vdjeric commented on sample-windows.ini:
```diff
@@ -0,0 +1,45 @@
+
+maxFileSize = 1048576
+
+logPath = %(TEMP)s\log\snappy
```
do we create the log directory ourselves? My Windows machine doesn't have %TEMP%\log\

vdjeric commented on sample-windows.ini:
```diff
@@ -0,0 +1,45 @@
+
+[SymbolPaths]
+
+Firefox = %(TEMP)s\breakpad\symbols_ffx
```
maybe %HOME% here as well

vdjeric commented on symbolicationWebService.py:
```diff
@@ -43,10 +47,27 @@
   ]
 }
 
+# Use a new class to make defaults case-sensite
```
typo in comment: sensitive

walac replied:
Ok, fixed.

vdjeric commented on symbolicationWebService.py:
```diff
@@ -146,50 +167,57 @@ def ReadConfigFile():
-      LogError("Unable to parse config file " + sys.argv[1] + ": " + str(e))
-    except:
-      LogError("Unable to open config file " + sys.argv[1])
+      logging.error("Unable to parse config file %s: %s", sys.argv[1], e)
```
why not LogError here?

walac replied:
symLogging has not been configured yet, so a call to `LogError` will raise an exception.

vdjeric replied:
ok

vdjeric commented on symbolicationWebService.py:
```diff
@@ -146,50 +167,57 @@ def ReadConfigFile():
-  configURLs = configParser.items("SymbolURLs")
-  if configURLs:
-    gOptions["symbolURLs"] = [url for name, url in configURLs]
+  if configParser.has_section("SymbolURLs"):
```
so you're going to allow having both SymbolPaths and SymbolURLs? yeah I guess that could be a valid config

vdjeric commented:
r+ with comments above addressed

vdjeric commented:
I'll merge it, and check what sample-linux.ini looks like

vdjeric commented:
sample-linux.ini has the extra newlines on my Windows machine, sample-windows.ini is missing it's LFs

vdjeric commented:
I fixed the newline problem with the last commit
